File: Cpp/897_Increasing_Order_Search_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* increasingBST(TreeNode* root) {
        return recursive(root, nullptr);
    }
    
    TreeNode* recursive(TreeNode* node, TreeNode* prev){
        if (!node) return prev;
        TreeNode *left = recursive(node->left, node);
        node->left = nullptr;
        node->right = recursive(node->right, prev);
        return left;
    }
};

File: Kotlin/897_Increasing_Order_Search_Tree.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun increasingBST(root: TreeNode?): TreeNode? {
        return recursive(root, null)
    }
    
    fun recursive(node: TreeNode?, prev: TreeNode?): TreeNode? {
        if(node == null) return prev
        
        val left = recursive(node.left, node)
        node.left = null
        node.right = recursive(node.right, prev)
        
        return left
    }
}